File: maxDepthOfBinaryTree.js
// https://leetcode.com/problems/maximum-depth-of-binary-tree/


const maxDepth = function(root) {
	return root ? 1 + Math.max(maxDepth(root.left), maxDepth(root.right)) : 0;
}

const TreeNode = function(val) {
	this.val = val;
	this.left = null;
	this.rigth = null;
}

console.assert(maxDepth(null) === 0, "works for a null node");
const test = new TreeNode(1);
console.assert(maxDepth(test) === 1, "works for a tree with a single node");
test.left = new TreeNode(2);
test.right = new TreeNode(3);
console.assert(maxDepth(test) === 2, "works for tree of depth 2");
test.left.left = new TreeNode(4);
test.left.right = new TreeNode(5);
console.assert(maxDepth(test) === 3, "works for tree of depth three");
test.left.right.left = new TreeNode(6);
test.left.right.left.right = new TreeNode(7);
console.assert(maxDepth(test) === 5, "works for a zig-zag depth");

File: maxDepthOfBinaryTree.py
# https://leetcode.com/problems/maximum-depth-of-binary-tree/

class TreeNode:
	def __init__(self, val = 0, left = None, right = None):
		self.val = val
		self.left = left
		self.right = right

class Solution:
	def maxDepth(self, root: TreeNode) -> int:
		if root == None: return 0
		depth = 1
		return depth + max(self.maxDepth(root.left), self.maxDepth(root.right))

test = Solution()
case1 = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))

print(test.maxDepth(case1))
